### PR TITLE
Implement gRPC based initial state settling in alertmanager.

### PR DESCRIPTION
**Pull request by stevesg**
**What this PR does**:
Implements setting the alertmanager state by obtaining copies of the full state from replicas over gRPC. This is only used when the new "Sharding" mode of alertmanager is enabled, and is otherwise unused.

The work is broken up into commits, which can be broken out into separate PRs if preferred, as each change is tested independently with unit tests.

Depends on: #3958 
Fixes: #3927

**Checklist**
- [X] Tests updated
- [X] ~~Documentation added~~
- [X] ~~`CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`~~

### CODE REVIEW

**Comment by stevesg** [comment]
Rebase + updates from refactoring

**Comment by stevesg** [comment]
Rebase

**Comment by stevesg** [comment]
Complete - just looking into unit test failure.

File: pkg/alertmanager/alertmanager.go
Comment (pracucci) [inline pkg/alertmanager/alertmanager.go:377]: Check if `am.state` is actually a `*state`. If not, return `error`.

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1041]:
```suggestion
// ReadFullStateForUser attempts to read the full state from each replica for user. Note that it will try to obtain and return
```

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1042]:
```suggestion
// state from all replicas, but will consider it a success if state is obtained from at least a replica.
```

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1015]:
```suggestion
// getOtherReplicasForUser returns the addresses for each _other_ replica assigned to a user.
```

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1019]: You can remove this check. It's already done in `am.ring.Get()`.

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1030]: You can call `replicationSet.GetAddresses()` to get the list of addresses. You can add another function like `ReplicationSet.GetAddressesWithout(exclude ...string)` for your convenience.

Once you apply both comments, this function becomes so small that I believe you don't even need it and you can embed the few lines of code directly in `ReadFullStateForUser()`.

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1078]: [nit] You can remove all these `return nil` moving them after the `switch`.

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1209]: This is already logged in the caller. Should we log it here as well?

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1193]: We should return the actual error:

```suggestion
		return nil, err
```

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1197]: Do we need to block until the end of the call? I think you just need to protect the access to `am.alertmanagers[userID]`, so you should be able to unlock the line after that. Am I missing anything?

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:30]: Naming it `settleBackoff` may be more clear.

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:56]: [nit] Move the default to a global constant.

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:50]: I think this config would end up retrying for too long (5 retries, 10s max backoff, 15s timeout). Before discussing how to fine tune it, question is: why do we need retries at all?

I understand "everything fail all the time", but do we need retries also for the normal use case (eg. the other replica is not "ready" yet) or just on case of failures while fetching it? If retries are to cover the latter case, then maybe we don't need retries at all, considering the design is to fallback reading state from the storage (which will be one of the next things we'll build).

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:206]: [nit] Can simplify with a guard:

```suggestion
				level.Error(s.logger).Log("msg", "key not found while merging full state", "user", s.userID, "key", p.Key)
				continue
```

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:207]: Isn't this a breaking error? Shouldn't we consider the settlement failed in case we can't merge the data (for any reason)?

I would return error from this function in the case, and consider the settlement failed in the caller. This will be important once we'll add the feature to load the state from storage (this is one of the cases where we should fallback to storage, I think).

File: pkg/alertmanager/state_replication_test.go
Comment (pracucci) [inline pkg/alertmanager/state_replication_test.go:70]: Should we return error given it's an unexpected condition in tests (eg. test fixture has not been mocked correctly)?

File: pkg/alertmanager/state_replication_test.go
Comment (pracucci) [inline pkg/alertmanager/state_replication_test.go:369]: [nit] You don't need sorting if you use `assert.ElementsMatch()` instead of `assert.Equal()`.

File: pkg/alertmanager/alertmanager.go
Comment (pracucci) [inline pkg/alertmanager/alertmanager.go:204]: We don't wait until started here (and it's correct). However, this means that we may start using this `Alertmanager` instance before settlement is completed (and I believe this is not correct). Am I missing anything?

**Review by pracucci — COMMENTED** [review]
Very good job! The logic flow is pretty clear. I left few nits.

My only question is whether we would end up in a situation where we could use the alertmanager before settlement is done (see comment in `Alertmanager.New()`).

Also, if initial state settlement works correctly, you should be able to remove the `t.Skip()` from the integration test `TestAlertmanagerSharding` and mention this PR fixes #3927, correct?

File: pkg/alertmanager/alertmanager.go
Comment (stevesg) [reply pkg/alertmanager/alertmanager.go:377]: :+1: Fix for the line above https://github.com/cortexproject/cortex/pull/3970

File: pkg/alertmanager/alertmanager.go
Comment (stevesg) [reply pkg/alertmanager/alertmanager.go:204]: I have been assuming we want to settle in the background, because an `Alertmanager` is spun up for every tenant. If they all have to hit the timeout, that might take too long if done in serial. That being said, it's harder to reason about correctness in this case

Perhaps safer to change it to wait for now, and explore doing it in the background as a separate piece of work?

File: pkg/alertmanager/multitenant.go
Comment (stevesg) [reply pkg/alertmanager/multitenant.go:1209]: Good question. It's often useful to see failed requests on the server side too, but perhaps a `Debug` level message is good enough, to keep the verbosity down?

File: pkg/alertmanager/state_replication.go
Comment (stevesg) [reply pkg/alertmanager/state_replication.go:50]: I don't really have a good answer for this, other than _that was the plan_. Maybe @gotjosh has an opinion?

I suppose we want to try and cover transient errors as best as possible? But given we're using a more reliable means of fetching the state, which itself has a timeout, and as you say, we have another fallback to S3, maybe we can take the retries away entirely.

File: pkg/alertmanager/state_replication.go
Comment (stevesg) [reply pkg/alertmanager/state_replication.go:207]: Agreed. Looking at the code for [silences](https://github.com/prometheus/alertmanager/blob/master/silence/silence.go#L821) and [notifications](https://github.com/prometheus/alertmanager/blob/master/nflog/nflog.go#L526), it seems the only cause of this error (currently) is failure to decode the message from the wire. Makes sense to fall back to the persisted state, if available.

I'm actually think if we should do the same for `key not found`?

File: pkg/alertmanager/state_replication_test.go
Comment (stevesg) [reply pkg/alertmanager/state_replication_test.go:369]: I'm checking equality the whole gRPC message, that seemed cleaner in case `FullState` ever gained new fields. I could compare just `Parts` instead? No strong opinion here really.

File: pkg/alertmanager/multitenant.go
Comment (stevesg) [reply pkg/alertmanager/multitenant.go:1197]: Have to admit I have not thought about it  - I was following on from the `UpdateState` implementation. I'll take a closer look to see if the mutex is protecting us from anything other than the map access.

**Review by stevesg — COMMENTED** [review]
Thanks for the review!

Yes I'm just experimenting with that test today to see if it's reliable enough to enable now.

**Comment by stevesg** [comment]
Rebase

File: pkg/alertmanager/multitenant.go
Comment (stevesg) [reply pkg/alertmanager/multitenant.go:1042]: I see the mistake, but changed to `at least one` replica instead.

**Comment by stevesg** [comment]
Addressed review comments.

File: pkg/alertmanager/multitenant.go
Comment (stevesg) [reply pkg/alertmanager/multitenant.go]: Same fix for `UpdateState`: https://github.com/cortexproject/cortex/pull/3972

File: pkg/alertmanager/alertmanager.go
Comment (stevesg) [inline pkg/alertmanager/alertmanager.go:380]: Going to update this message to match https://github.com/cortexproject/cortex/commit/c34eb640df12bd6e9ff40dee057696184ad03dc8

**Comment by stevesg** [comment]
Rebase

**Comment by stevesg** [comment]
I've done some testing locally on https://github.com/cortexproject/cortex/issues/3927 and before this change it failed 3/20, and afterwards 0/20. So looking good but I want to check through the logs to check we're working as expected, then run a few more iterations before re-enabling the test.

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:204]: It's worth noting that "start using" is not as one would think. Yes, we'll accept alerts, silences, etc. However, we'll wait for the state to be replicated before we send a notification.

File: pkg/alertmanager/state_replication_test.go
Comment (pracucci) [reply pkg/alertmanager/state_replication_test.go:370]: Ok got it. That what you do is fine 👍

File: pkg/ring/replication_set.go
Comment (pracucci) [inline pkg/ring/replication_set.go:119]: [nit] 
```suggestion
// excluding the specified addresses. Returned slice order is not guaranteed.
```

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1074]: I find this comment deceiving for a couple of reasons:
1. It's not this function retrying but who calls it. This function should know nothing about who calls it.
2. I'm not even sure the caller retries in this case. Does it?

Zooming out, why isn't this returned as an error? It means we failed to sync the state from other replicas (eg. think about an alertmanagers cluster coldstart), so shouldn't we try to read from storage (which is the next step of your work)?

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:171]: From a previous comment. Steve wrote:

> I suppose we want to try and cover transient errors as best as possible? But given we're using a more reliable means of fetching the state, which itself has a timeout, and as you say, we have another fallback to S3, maybe we can take the retries away entirely.

I'm leaning towards remove retries entirely, unless we have a good reason to keep it, because of the reasons already mentioned.

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:213]: From a previous comment. Steve wrote:

> Looking at the code for silences and notifications, it seems the only cause of this error (currently) is failure to decode the message from the wire. Makes sense to fall back to the persisted state, if available.
> I'm actually think if we should do the same for key not found?

If a key has not been found it means that key hasn't been added by `state.AddState()`, which means this alertmanager version doesn't know what to do with that key (eg. the key may have been added by a newer version of the alertmanager and a mix of old and new versions are cohexisting for a while). In this case, I would continue to log but not returning an error.

File: pkg/alertmanager/multitenant_test.go
Comment (pracucci) [inline pkg/alertmanager/multitenant_test.go:1272]: [nit] 
```suggestion
					require.NoError(t, services.StopAndAwaitTerminated(ctx, am))
```

**Review by pracucci — COMMENTED** [review]
Very good job! LGTM (modulo last comments). Can you remove the skip from the integration test `TestAlertmanagerSharding` so we make sure it passes?

File: pkg/alertmanager/state_replication.go
Comment (stevesg) [reply pkg/alertmanager/state_replication.go:171]: @gotjosh

File: pkg/alertmanager/multitenant.go
Comment (stevesg) [reply pkg/alertmanager/multitenant.go]: Removed - no need for it without retry logic.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go]: The concept of retries in this scenario is more about making sure we've communicated with other replicas than it is about recovering from failures. 

What we care about here is two things:
- Make sure we've communicated with other replicas (if any)
- If no state was available from other replicas then try object storage

On scale up / downs retries might not matter all that much, but on total cluster failure or cluster startup I feel like we do care about retrying because other replicas might also starting up.

I guess the main thing to address as part of this logic IMO is that we should keep track of whenever `len(fullStates) == replication factor` and if is we should carry on.

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [inline pkg/alertmanager/alertmanager.go:132]:
```suggestion
	// ReadFullStateForUser obtains the full state from other replicas in the cluster.
```

Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:132]: (sorry for the nit, just to keep the language consistent with others above)

File: pkg/ring/replication_set.go
Comment (gotjosh) [inline pkg/ring/replication_set.go:119]:
```suggestion
// excluding the specified address. Returned slice order is not guaranteed.
```

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [inline pkg/alertmanager/state_replication.go:229]: Do we need this? IIRC, we decided to go with a no rebroadcasting of state on calls to `Merge`  under that scheme this would never occur unless we received a request directly to this Alertmanager

**Review by gotjosh — COMMENTED** [review]
LGTM Steve 👏, I have one question on the readiness and re-broadcasting and I think we need a decision on the semantics of "retries".

File: pkg/alertmanager/state_replication.go
Comment (stevesg) [reply pkg/alertmanager/state_replication.go:229]: Good spot - yes we do need it for now, until the re-broadcasting is disabled. I wanted to do it before this commit, but it seems that disabling it is not trivial as we thought and will involve upstream changes.

File: pkg/alertmanager/alertmanager.go
Comment (stevesg) [reply pkg/alertmanager/alertmanager.go:204]: Will leave as-is. My (current) understanding is as Josh said - there is no requirement to block (except for notifications, which are blocked via the call into `WaitReady`).

**Review by gotjosh — APPROVED** [review]
LGTM

**Review by pracucci — APPROVED** [review]
Thanks for addressing all comments. LGTM! 🚀

**Review by pstibrany — APPROVED** [review]

**Pull request merged**
